### PR TITLE
Fix : URL Alternate with unwanted parameters

### LINKED ISSUE #28609: URL Alternate with unwanted parameters
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Hello,

I have just discovered an embarrassing SEO bug, which multiplies the number of existing URLs for a page.

This bug concerns the ModuleFrontController (and function hookModuleRoutes) with several languages which are activated on the Prestashop. 

When there is a well-declared rewrite at the module level, such as for the SmartBlog module, the URL generates well with the URL rewrite. However, when Prestashop generates "alternate" URLs, it goes back to the "sanitizeUrl" function, which adds the parameters that are supposed to be hidden.

Here is an example :
```html
<link rel="alternate" hreflang="fr-fr" href="http://www.test.local/fr/blog/category/15_des-conseils.html?slug=des-conseils&module=smartblog">
```

### Expected behavior

We should have :
```html
<link rel="alternate" hreflang="fr-fr" href="http://www.test.local/fr/blog/category/15_des-conseils.html">
```

Checking the code:
https://github.com/PrestaShop/PrestaShop/blob/531c7407104c4d7cfc72c28808ea9fa80794ff7b/classes/controller/FrontController.php#L2004
We can see that the Sanitize is not useful, because the URL is already generated by Prestashop upstream:
https://github.com/PrestaShop/PrestaShop/blob/531c7407104c4d7cfc72c28808ea9fa80794ff7b/classes/controller/FrontController.php#L1995

I therefore propose that this sanitize be removed unless there are a few things that escape me.

For my part, I fixed it with an override from a module:
`override/classes/controller/FrontController.php`
```php
<?php

class FrontController extends FrontControllerCore
{
    /**
     * @return array containing the URLs of the same page but for different languages
     */
    protected function getAlternativeLangsUrl()
    {
        $alternativeLangs = [];
        $languages = Language::getLanguages(true, $this->context->shop->id);
        if (count($languages) < 2) {
            return $alternativeLangs;
        }
        foreach ($languages as $lang) {
            $langUrl = $this->context->link->getLanguageLink($lang['id_lang']);
            $alternativeLangs[$lang['language_code']] = $langUrl;
        }
        return $alternativeLangs;
    }
}
```

### Steps to reproduce

1. Add the SmartBlog module (I think this is a problem with all modules that use ModuleFrontController and the hookModuleRoutes hook)
2. Add at least two languages
3. Create a blog category
4. Check on the category page, Front side, the "alternate" tag

### PrestaShop version(s) where the bug happened

1.7.8.5

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

https://github.com/smartdatasoft/smartblog

### PR DESCRIPTION
**Pull request by tups**
<table width="100%" style="display: table;table-layout: fixed;width: 100%;">
<thead>
<tr>
<th style="width: 18%;">Questions</th>
<th>Answers</th>
</tr>
</thead>
<tbody>
<tr>
<td>Branch?</td>
<td>develop</td>
</tr>
<tr>
<td>Description?</td>
<td>
I have just discovered an embarrassing SEO bug, which multiplies the number of existing URLs for a page.<br><br>This bug concerns the ModuleFrontController (and function hookModuleRoutes) with several languages which are activated on the Prestashop. <br><br>When there is a well-declared rewrite at the module level, such as for the example module, the URL generates well with the URL rewrite. However, when Prestashop generates "alternative" URLs, it executes the "sanitizeUrl" function, which adds parameters that are supposed to be hidden.<br><br>I therefore propose that this sanitize be removed unless there are a few things that escape me.
</td>
</tr>
<tr>
<td>Type?</td>
<td>bug fix</td>
</tr>
<tr>
<td>Category?</td>
<td>FO</td>
</tr>
<tr>
<td>BC breaks?</td>
<td>no</td>
</tr>
<tr>
<td>Deprecations?</td>
<td>no</td>
</tr>
<tr>
<td>Fixed ticket?</td>
<td>Fixes <a class="issue-link js-issue-link" data-error-text="Failed to load title" data-id="1250480241" data-permission-text="Title is private" data-url="https://github.com/PrestaShop/PrestaShop/issues/28609" data-hovercard-type="issue" data-hovercard-url="/PrestaShop/PrestaShop/issues/28609/hovercard" href="https://github.com/PrestaShop/PrestaShop/issues/28609">#28609</a> .</td>
</tr>
<tr>
<td>Related PRs</td>
<td>If theme, autoupgrade or other module change is needed, provide a link to related PRs here.</td>
</tr>
<tr>
<td>How to test?</td>
<td>To illustrate the bug, as precisely as possible, I have just developed a specific module for this issue.<br>A simple module, with a Front Controller.<br><br>It is therefore sufficient:<br>- to set up at least two languages on the Prestashop, for example French and English<br>- To install the module:<br><a href="https://github.com/PrestaShop/PrestaShop/files/8787141/lowissuealternate.zip">lowissuealternate.zip</a><br>- In the configuration of the module, a button allows access to the URL of the module on the Front side, you will see the "alternate" tags<br>- You will be able to see that it has nothing to do with the URL of the page in your web browser<br>- You can also check that it does not come from the module, but from the function that generates the "alternate"
</td>
</tr>
<tr>
<td>Possible impacts?</td>
<td>No</td>
</tr>
</tbody>
</table>

### Here is an example :

```html
<link rel="alternate" hreflang="fr-fr" href="http://www.test.local/fr/blog/category/15_des-conseils.html?slug=des-conseils&module=smartblog">
```
### Expected behavior

We should have :
```html
<link rel="alternate" hreflang="fr-fr" href="http://www.test.local/fr/blog/category/15_des-conseils.html">
```

Checking the code:
https://github.com/PrestaShop/PrestaShop/blob/531c7407104c4d7cfc72c28808ea9fa80794ff7b/classes/controller/FrontController.php#L2004
We can see that the Sanitize is not useful, because the URL is already generated by Prestashop upstream:
https://github.com/PrestaShop/PrestaShop/blob/531c7407104c4d7cfc72c28808ea9fa80794ff7b/classes/controller/FrontController.php#L1995

I therefore propose that this sanitize be removed unless there are a few things that escape me.

### CODE REVIEW

**Review by Hlavtox — APPROVED** [review]
Everything OK from SEO perspective. 👍

**Comment by ghost** [comment]
Nice cacth, but i think you need to target develop branch ?

Edit : oups it's good just the description of the PR to change

**Review by ghost — APPROVED** [review]

**Comment by PaRomainCouderc** [comment]
There is the same thing for canonicalRedirection method.
PrestaShop do redirection with parameters that are supposed to be hidden.

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
I've just approved a PR, and I don't think it should be validated by a PM but by a dev.

@Hlavtox, could you comment on the issue? To me, it looks like a legitimate change.

**Comment by kpodemski** [comment]
ping @PrestaShop/product-team

**Comment by Hlavtox** [comment]
@kpodemski @tups @okom3pom The change is definitely correct from logical point of view, but I think the main problem is in `sanitizeUrl` function. Why the hell should "Sanitize / Clean params of an URL" add some new parameters. 🤔

**Comment by Hlavtox** [comment]
I investigated some more, you can reproduce with this module also.
https://github.com/PrestaShop/PrestaShop/files/9293465/tox_layouttest.zip

If we add `'module'` to `$excluded_key` array in FrontController::sanitizeUrl()... it fixes the friendly URL scenario.

### Friendly URLs enabled
Go to /cs/layouttest
You get correct URLs
`http://localhost/dev/cs/layouttest`
`http://localhost/dev/gb/layouttest`

### Friendly URLs disabled
This is broken even after the fix, but in a different way. 
Go to http://localhost/dev/cs/index.php?fc=module&module=tox_layouttest&controller=layouttest

getLanguageLink returns:
`http://localhost/dev/index.php?id_lang=1&fc=module&module=tox_layouttest&controller=layouttest&id_lang=1`
`http://localhost/dev/index.php?id_lang=2&fc=module&module=tox_layouttest&controller=layouttest&id_lang=2`

after sanitizeURL:
`http://localhost/dev/index.php?id_lang=1&fc=module&module=tox_layouttest&controller=layouttest`
`http://localhost/dev/index.php?id_lang=2&fc=module&module=tox_layouttest&controller=layouttest`

you can also access it by and not get redirected anywhere:
`http://localhost/dev/index.php?id_lang=1&fc=module&module=tox_layouttest&controller=layouttest&id_lang=1`
`http://localhost/dev/index.php?id_lang=1&fc=module&module=tox_layouttest&controller=layouttest`

-----------------------------------
TLDR:
- Prestashop MUST redirect to one unique canoncial URL.
- This one single canoncial URL must be present in HREF tags.
- If additional parameters not belonging to a core are in the URL, they must be kept in HREF tags.

**Comment by MatShir** [comment]
So there are still some improvement to add in this PR @Hlavtox ? Should I remove the waiting for QA ?

**Review by MhiriFaten — COMMENTED** [review]
Hello @tups ,

I tested the PR and I noticed the following issues:

**Issue 1**

The slug is still visible in the URL not like mentioned in the expected behavior

image

**Issue 2**

1. Go to configure
2. Click on `SHOW MODULE CONTROLLER FRONT` button
3. set the URL
4. see error 

[Untitled_ Nov 11, 2022 4_01 PM.webm](https://user-images.githubusercontent.com/111756615/201367224-de1fa5a5-2b77-4bce-af55-0c476f9b38c6.webm)

**Issue 3**

1. Go to BO
2. Shop Parameters > Traffic & SEO
3. Enable `Friendly URL`
4. See error -> the slug and module still visible in the URL

image

Could you please check? :pray: 

Thank you,

**Comment by Hlavtox** [comment]
@MatShir @kpodemski I don't think it's a correct way to fix this. We should fix the `sanitizeUrl` function I think.

**Comment by kpodemski** [comment]
Since we have had no news from you for more than 20 days, I'm closing this PR. Feel free to re-open it if you find the motivation to finish it. Don't hesitate to ping me if you have any questions or doubts.
